src/background.ts

```diff
@@ -0,0 +1,287 @@
+        rallyId = "11111111-1111-1111-1111-111111111111";
+      }
+      console.info(`Study running with Rally ID: ${rallyId}`);
+
+      const storage = await browser.storage.local.get("enrolled");
+      if (storage.enrolled !== true) {
+        console.info("Recording enrollment.");
+        rallyManagementMetrics.id.set(rallyId);
+        examplePings.studyEnrollment.submit();
+
+        browser.storage.local.set({
+          enrolled: true,
+        });
+      }
+
+      // The Rally API has been initialized.
+      // Initialize the study and start it.
+
+      // Example: initialize the example module.
+      exampleModuleInitialize();
+      await browser.storage.local.set({ "state": RunStates.Running });
+
+
+      // Example: set a listener for WebScience page navigation events on
+      // http://localhost/* pages. Note that the manifest origin
+      // permissions currently only include http://localhost/*. You should
+      // update the manifest permissions as needed for your study.
+
+      this.pageDataListener = async (pageData) => {
+        console.log(`WebScience page navigation event fired with page data:`, pageData);
+        userJourney.pageId.set(pageData.pageId);
+        // Round attention duration up to 1ms, Glean will balk if it is lower than this.
+        if (pageData.attentionDuration < 1) {
+          userJourney.attentionDuration.set(1);
+        } else {
+          userJourney.attentionDuration.set(pageData.attentionDuration);
+
+        }
+        userJourney.audioDuration.set(pageData.audioDuration);
+        // Max relative scroll depth is a percentage expressed as a decimal by WebScience,
+        // but Glean is expecting an integer.
+        userJourney.maxRelativeScrollDepth.set(Math.floor(pageData.maxRelativeScrollDepth * 100));
+
+        const pageVisitStart = new Date(pageData.pageVisitStartTime);
+        const pageVisitStop = new Date(pageData.pageVisitStopTime);
+        userJourney.pageVisitStartDateTime.set(pageVisitStart);
+        userJourney.pageVisitStopDateTime.set(pageVisitStop);
+        // Referrer is optional, and will be an empty string if unset.
+        if (pageData.referrer) {
+          userJourney.referrer.setUrl(pageData.referrer);
+        }
+        userJourney.url.setUrl(pageData.url);
+
+        // Submit the metrics constructed above.
+        examplePings.userJourney.submit();
+      };
+
+      webScience.pageNavigation.onPageData.addListener(this.pageDataListener, { matchPatterns: ["<all_urls>"] });
+
+      // Example: register a content script for http://localhost/* pages
+      // Note that the content script has the same relative path in dist/
+      // that it has in src/. The content script can include module
+      // dependencies (either your own modules or modules from npm), and
+      // they will be automatically bundled into the content script by
+      // the build system.
+      this.contentScript = webScience.contentScripts.registerContentScript(
```

Comment:
Note: we need to update WebScience so that it exports this function. It's a cross-browser wrapper around dynamic content script loading, which Chrome currently doesn't support (so for Chrome, the content scripts get built and put in dist/ and a message gets logged to the console telling you that you need to load them in the manifest)

Reply:
https://github.com/mozilla-rally/web-science/pull/278